From 39b2b1b6276ea02cc9f6e2c54c71abe974e25a8d Mon Sep 17 00:00:00 2001
From: DEREFERENC3D <dereferenc3d@protonmail.com>
Date: Fri, 11 Oct 2024 21:42:50 +0200
Subject: [PATCH 1/1] Support setting base URLs for every part of Piped

* Allow overriding the addresses of other parts of Piped than just the
  backend,
* Make the values full URLs instead of hostnames, removing the
  requirement of a separate subdomain for every server app that Piped
  consists of,
* Allow setting the related environment variables at runtime
  instead of at build time.
---
 .env                 |  3 ---
 Dockerfile           |  7 ++++++-
 docker/entrypoint.sh | 44 +++++++++++++++++++++++++++++++++-----------
 localizefonts.sh     |  2 +-
 vite.config.js       |  1 +
 5 files changed, 41 insertions(+), 16 deletions(-)
 delete mode 100644 .env

diff --git a/.env b/.env
deleted file mode 100644
index ddba8b7a..00000000
--- a/.env
+++ /dev/null
@@ -1,3 +0,0 @@
-VITE_PIPED_API=https://pipedapi.kavin.rocks
-VITE_PIPED_PROXY=https://pipedproxy.kavin.rocks
-VITE_PIPED_INSTANCES=https://piped-instances.kavin.rocks/
diff --git a/Dockerfile b/Dockerfile
index 8115599c..804d0a23 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -13,7 +13,12 @@ RUN corepack enable && corepack prepare pnpm@latest --activate
 RUN --mount=type=cache,target=/root/.local/share/pnpm \
     --mount=type=cache,target=/app/node_modules \
     pnpm install --prefer-offline && \
-    pnpm build && ./localizefonts.sh
+    env \
+        VITE_PIPED_API=API_URL_GOES_HERE \
+        VITE_PIPED_PROXY=PROXY_URL_GOES_HERE \
+        VITE_PIPED_INSTANCES=INSTANCES_URL_GOES_HERE \
+    pnpm build && \
+    chmod +x ./localizefonts.sh && ./localizefonts.sh
 
 FROM nginxinc/nginx-unprivileged:alpine
 
diff --git a/docker/entrypoint.sh b/docker/entrypoint.sh
index b17a2253..17d73743 100755
--- a/docker/entrypoint.sh
+++ b/docker/entrypoint.sh
@@ -1,21 +1,43 @@
 #!/bin/sh
 
-if [ -z "${BACKEND_HOSTNAME}" ]; then
-    echo "BACKEND_HOSTNAME not set"
-    exit 1
-fi
+SITE='/usr/share/nginx/html'
+ASSETS="${SITE}/assets"
 
-HTTP_MODE=${HTTP_MODE:-https}
+# Set the URLs for every component of Piped
+# base URL for the frontend
+if [ -z "${BASE_URL}" ]; then
+	echo "BASE_URL not set"
+	exit 1
+fi
+sed -i "s|/BASE_URL_GOES_HERE|${BASE_URL}|" "${SITE}/index.html"
+# I have no idea why but 'sed' does not get everything at once, only in this case...
+while grep -q '/BASE_URL_GOES_HERE' ${ASSETS}/*; do
+	sed -i "s|/BASE_URL_GOES_HERE|${BASE_URL}|" ${ASSETS}/*
+done
+# second occurence - wihtout the leading slash
+while grep -q 'BASE_URL_GOES_HERE' ${ASSETS}/*; do
+	sed -i "s|BASE_URL_GOES_HERE|${BASE_URL}|" ${ASSETS}/*
+done
 
-sed -i "s@https://pipedapi.kavin.rocks@${HTTP_MODE}://pipedapi.kavin.rocks@g" /usr/share/nginx/html/assets/*
-sed -i "s/pipedapi.kavin.rocks/${BACKEND_HOSTNAME}/g" /usr/share/nginx/html/assets/*
+# set the backend's URL
+if [ -z "${BACKEND_URL}" ]; then
+	echo "BACKEND_URL not set"
+	exit 1
+fi
+sed -i "s|API_URL_GOES_HERE|${BACKEND_URL}|" ${ASSETS}/*
 
-if [ -n "${HTTP_WORKERS}" ]; then
-    sed -i "s/worker_processes  auto;/worker_processes  ${HTTP_WORKERS};/g" /etc/nginx/nginx.conf
+# proxy's URL
+if [ -z "${PROXY_URL}" ]; then
+	echo "PROXY_URL not set"
+	exit 1
 fi
+sed -i "s|PROXY_URL_GOES_HERE|${PROXY_URL}|" ${ASSETS}/*
 
-if [ -n "${HTTP_PORT}" ]; then
-    sed -i "s/80;/${HTTP_PORT};/g" /etc/nginx/conf.d/default.conf
+# instance list
+if [ -z "${INSTANCES_URL}" ]; then
+	echo "INSTANCES_URL not set"
+	exit 1
 fi
+sed -i "s|INSTANCES_URL_GOES_HERE|${INSTANCES_URL}|" ${ASSETS}/*
 
 nginx -g "daemon off;"
diff --git a/localizefonts.sh b/localizefonts.sh
index 45a2c366..420364aa 100755
--- a/localizefonts.sh
+++ b/localizefonts.sh
@@ -8,4 +8,4 @@ for font in $fonts; do
 	mkdir -p "$(dirname "$file")"
 	curl -L "$font" -o "$file"
 done
-sed -Ei "s#$base#/fonts#g" dist/assets/*
+sed -Ei "s#$base#BASE_URL_GOES_HERE/fonts#g" dist/assets/*
diff --git a/vite.config.js b/vite.config.js
index 3b455b09..8b27f0df 100644
--- a/vite.config.js
+++ b/vite.config.js
@@ -9,6 +9,7 @@ import eslintPlugin from "vite-plugin-eslint";
 
 // https://vitejs.dev/config/
 export default defineConfig({
+    base: 'BASE_URL_GOES_HERE',
     plugins: [
         vue(),
         Unocss(),
-- 
2.49.0

